### PR TITLE
Empowered BlockProcessor to be solely responsible for block processing

### PR DESCRIPTION
**Pull request by osmirnov**
* Empowered BlockProcessor to be solely responsible for block processing
* Tweaked Pool class to use Dictionary
* Switched handler to interact with new BlockProcessor

### CODE REVIEW

File: src/NeoSharp.Core/Blockchain/Blockchain.cs [hunk 1]
```diff
@@ -109,93 +119,20 @@ static int TransactionComparer(Stamp<Transaction> a, Stamp<Transaction> b)
...
-                return Task.FromResult(false);
-            }
-
-            return Task.FromResult(BlockPool.Push(block));
+            _blockProcessor.AddBlock(block);
         }
 
         /// <inheritdoc />
-        public async Task<bool> PersistBlock()
+        public Task<bool> ContainsBlock(UInt256 blockHash)
```
Comment (aboimpinto) [inline src/NeoSharp.Core/Blockchain/Blockchain.cs]: This method is not necessary. You can just call the ContainsBlock from anywhere in the BlockChain class.
Even the validations can be added to the BlockProcessor.ContainsBlock.

File: src/NeoSharp.Core/Blockchain/Blockchain.cs [hunk 2]
```diff
@@ -67,6 +69,27 @@ public Blockchain(IRepository repository, IBinarySerializer serializer, ICrypto
                 return;
             }
 
+            _blockProcessor.OnBlockProcessed += async (_, block) =>
```
Comment (aboimpinto) [inline src/NeoSharp.Core/Blockchain/Blockchain.cs]: why add an inline anonymous method and not a real method?

File: src/NeoSharp.Core/Blockchain/Blockchain.cs [hunk 3]
```diff
@@ -109,93 +119,20 @@ static int TransactionComparer(Stamp<Transaction> a, Stamp<Transaction> b)
         #region Blocks
 
         /// <inheritdoc />
-        public Task<bool> AddBlock(Block block)
+        public void AddBlock(Block block)
```
Comment (aboimpinto) [inline src/NeoSharp.Core/Blockchain/Blockchain.cs]: another method that I see no reason to live in BlockChain class and IBlockChain interface. Who need this in the rest of the application should use the IBlockProcessor interface.

**Comment by aboimpinto** [comment]
I think the BlockProcessor is a great name and has a great goal and remove some of the dependencies from BlockChain class.

**Review by aboimpinto — APPROVED** [review]

File: src/NeoSharp.Core/Blockchain/Blockchain.cs (hunk 2 above)
Comment (osmirnov) [reply src/NeoSharp.Core/Blockchain/Blockchain.cs]: Yes I will turn it into the real method

**Comment by osmirnov** [comment]
@aboimpinto you are absolutely right about AddBlock/ContainsBlock on Blockchain class. Based on my perception we can drop them and switch to BlockProcessor class instead.

**Pull request merged**